Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -336,6 +336,34 @@ instance : Coe (FiniteAdeleRing R K) (K_hat R K) where
 lemma ext {a₁ a₂ : FiniteAdeleRing R K} (h : (a₁ : K_hat R K) = a₂) : a₁ = a₂ :=
   Subtype.ext h
 
+@[simp]
+lemma mem_FiniteAdeleRing (x : K_hat R K) : x ∈ (
+    { carrier := {x : K_hat R K | x.IsFiniteAdele}
+      mul_mem' := mul
+      one_mem' := one
+      add_mem' := add
+      zero_mem' := zero
+      algebraMap_mem' := algebraMap'
+    } : Subalgebra K (K_hat R K)) ↔ {v | x v ∉ adicCompletionIntegers K v}.Finite := Iff.rfl
```

Comment:
I think the subalgebra should be separated into a new def, and `FiniteAdeleRing` could potentially be defined as `{ x : K_hat R K // x.IsFiniteAdele }` and only view it as a `subalgebra` (since they are defeq) when needed.

Reply:
I'd just remove the coercion in FiniteAdeleRing and change its type to Subalgebra.

Reply:
The problem with that is that `K_hat R K`, the thing it's a subalgebra of, is a technical intermediate thing, but `FiniteAdeleRing R K` is the fundamental definition and I think it deserves to be a type. There's a problem with having coerced terms as fundamental objects, and that's that typeclass inference goes crazy trying to deal with it. Here's the issue. We need to let `FiniteAdeleRing` be a type because it has, for example, a topology on it (and it's not the subspace topology coming from K_hat). So if it's a subalgebra then the code will be full of `↥FiniteAdeleRing`, and when typeclass sees that it wants e.g. a topological space structure on this, or a coercion from this to somewhere else, it is quick to start looking looking at the auxiliary object `K_hat` for answers. This leads to timeouts (we saw it with `↥(𝓞 K)` in the theory of number fields) and this is exactly why we changed the definition of ring of integers from a term to a type.

Reply:
I'm going to try erdOne's suggestion.